### ISSUE TITLE
Bug in CustomIntegrator for CUDA and OpenCL platforms

### ISSUE THREAD

**Issue by craabreu** [openmm/openmm#3183]
The script below works nicely in the `Reference` and `CPU` platforms, but fails in the `CUDA` and `OpenCL` ones.

```py
from simtk import openmm, unit
from simtk.openmm import app
from sys import stdout

platform = openmm.Platform.getPlatformByName('CUDA')

force_field = app.ForceField('tip3p.xml')
modeller = app.Modeller(app.Topology(), [])
modeller.addSolvent(force_field, boxSize=2.5*openmm.Vec3(1, 1, 1))
system = force_field.createSystem(modeller.topology, rigidWater=False)

tau = 100*unit.femtoseconds
T = 300*unit.kelvin
kB = unit.MOLAR_GAS_CONSTANT_R

# Massive, LFMiddle-type, Generalized Gaussian Moment Thermostat

integrator = openmm.CustomIntegrator(1*unit.femtosecond)
integrator.addGlobalVariable('kT', kB*T)
integrator.addGlobalVariable('Q1', kB*T*tau**2)
integrator.addGlobalVariable('Q2', 8/3*(kB*T)**3*tau**2)
integrator.addPerDofVariable('v1', 0)
integrator.addPerDofVariable('v2', 0)
integrator.addComputePerDof('v', 'v + dt*f/m')
integrator.addComputePerDof('x', 'x + 0.5*dt*v')
integrator.addComputePerDof('v1', 'v1 + 0.5*dt*(m*v^2 - kT)/Q1')
integrator.addComputePerDof('v2', 'v2 + 0.5*dt*(m^2*v^4/3 - kT^2)/Q2')
integrator.addComputePerDof('v', 'v*exp(-dt*(v1 + kT*v2))/sqrt(1 + 2*dt*m*v^2*v2/3)')
integrator.addComputePerDof('v2', 'v2 + 0.5*dt*(m^2*v^4/3 - kT^2)/Q2')
integrator.addComputePerDof('v1', 'v1 + 0.5*dt*(m*v^2 - kT)/Q1')
integrator.addComputePerDof('x', 'x + 0.5*dt*v')

simulation = app.Simulation(modeller.topology, system, integrator, platform)
simulation.context.setPositions(modeller.positions)
simulation.context.setVelocitiesToTemperature(T)
simulation.reporters.append(
    app.StateDataReporter(stdout, 10, step=True, potentialEnergy=True, temperature=True)
)
simulation.step(10000)
```

This is the error message when CUDA is used:

```
simtk.openmm.OpenMMException: Error compiling program: (774): error: no suitable constructor exists to convert from "float" to "double3"

(776): error: no suitable conversion function from "double3" to "real" exists

(779): error: no operator "=" matches these operands
            operand types are: double3 = real

(816): error: no suitable constructor exists to convert from "float" to "double3"

(818): error: no suitable conversion function from "double3" to "real" exists

(821): error: no operator "=" matches these operands
            operand types are: double3 = real

(754): warning: variable "gaussianIndex" was declared but never referenced

(755): warning: variable "uniformIndex" was declared but never referenced

6 errors detected in this compilation.
```

This is the one when OpenCL is used:

```
simtk.openmm.OpenMMException: Error compiling kernel: <kernel>:107:9: warning: 'WORK_GROUP_SIZE' macro redefined
#define WORK_GROUP_SIZE 512
        ^
<kernel>:24:9: note: previous definition is here
#define WORK_GROUP_SIZE 64
        ^
<kernel>:198:6: error: initializing '__attribute__((address_space(16776963))) real' (aka '__attribute__((address_space(16776963))) float') with an expression of incompatible type 'double3'
real multiplier = (make_double3(velocity.x, velocity.y, velocity.z));
     ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<kernel>:240:6: error: initializing '__attribute__((address_space(16776963))) real' (aka '__attribute__((address_space(16776963))) float') with an expression of incompatible type 'double3'
real multiplier = (make_double3(velocity.x, velocity.y, velocity.z));
     ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The issue goes away if we simply replace `v^4` by `(v^2)^2` in the following instruction (which appears twice in the script):

```
integrator.addComputePerDof('v2', 'v2 + 0.5*dt*(m^2*v^4/3 - kT^2)/Q2')
```

**Comment by peastman** [comment]
Could you try the latest nightly build?  You can install it with

```
conda install -c conda-forge -c omnia-dev openmm
```

There have been a lot of changes to that code, so the problem may have already been fixed.

**Comment by craabreu** [comment]
I tried the script with the latest nightly build. Unfortunately, both errors persist.

**Comment by peastman** [comment]
Fixed in #3187.  Thanks for the excellent test case!
